### PR TITLE
[SSHD-1050] Fixed race condition in AuthFuture if exception caught before authentication started

### PR DESCRIPTION
**Pull request by lgoldstein**

### CODE REVIEW

File: sshd-core/src/main/java/org/apache/sshd/client/session/ClientSessionImpl.java [hunk 1]
```diff
@@ -124,8 +123,19 @@ public AuthFuture auth() throws IOException {
...
+            Throwable caught = (authFuture instanceof InitialAuthFuture) ? authFuture.getException() : null;
             authFuture = ValidateUtils.checkNotNull(
                     authService.auth(serviceName), "No auth future generated by service=%s", serviceName);
+            if (caught != null) {
+                /*
+                 * Safe to do under session lock despite SSHD-916 since
+                 * the newly generated future has no associated listeners
+                 * attached to it yet
+                 */
+                signalAuthFailure(authFuture, caught);
```
[the remark this answers is not in the document]
Comment (lgoldstein) [reply sshd-core/src/main/java/org/apache/sshd/client/session/ClientSessionImpl.java:136]: Seems a good idea...

**Comment by tomaswolf** [comment]
I've pushed https://github.com/tomaswolf/mina-sshd/commit/3f15c6dbcce1dae0c88cae5b2ad0c2f07ebbc43e in https://github.com/tomaswolf/mina-sshd/tree/SSHD-1050 . It's a follow-up commit on this one; IMO solves the problem and I don't see any race conditions anymore. Feel free to fetch and squash into your change here.

Ran a full build of sshd, no tests failed. JGit tests also work with this.

BTW, I like your test for this; it reproduces consistently; much better than my original attempt.

**Comment by lgoldstein** [comment]
A bit complex to my taste - I have pushed a fix for my original commit - basically I am using an `AtomicReference` instead of `volatile` to manage the replacement of the current future instance. I believe this also solves the issue of multi-threading update of the authentication failure.

**Comment by tomaswolf** [comment]
Sorry, that doesn't change anything. The following sequence is still possible:

exceptionCaught: calls authFutureHolder.get(), gets the InitialAuthFuture
auth: calls prev = authFutureHolder.getAndSet(authFuture), gets the InitialAuthFuture
auth: calls prev.getException(), gets null
signalAuthFailure: calls setException() on the InitialAuthFuture.

The authFuture must be volatile (or an AtomicReference) not because of this race that's causing problems here, but because of the only partially synchronized accesses. To avoid the race, the _pairs_ of getting/setting the authFuture _and_ getting/setting the exception must be synchronized together. That was the point of my change.

**Comment by lgoldstein** [comment]
>> Sorry, that doesn't change anything. The following sequence is still possible

I agree - reached this conclusion myself shortly after proposing this change. Will look at your solution...

**Comment by lgoldstein** [comment]
Looks good - while reading it I was wondering what if the user calls `auth()` again (e.g., after unsuccessful attempt). In such a case, the `earlyErrors` have been nullified so on the 2nd call they will no longer be accumulated. According to the code in the `ClientUserAuthService` it is allowed to start a new `auth()` even while another is in progress. I therefore think that we should only save the first signaled error after a null in an `AtomicReference` and nullify its contents when new auth occurs.

I have pushed your changes + the above change for your review. It does pose an interesting question which I believe is not covered by both our changes: since the early error is nullified, a 2nd attempt will not "sense" it... Is it OK to have this "gap" and rely on the user timing out ? If so, the how do we prevent the user from using the infinite `verify` or `await` ?

**Comment by lgoldstein** [comment]
I have also pushed a change for multiple authentication attempts and added/modified `ClientSessionTest` accordingly - I would appreciate your feedback.

**Comment by tomaswolf** [comment]
I've been wondering about that, too. But I have no use case for calling auth() more than once on a session, and it's unclear to me what that would even mean. Or calling auth() without somehow checking the returned future.  But in any case, I don't think there's a gap. (Unless the client code really doesn't check the future from the first auth().)  After the first call to auth(), there will always be an AuthFuture.

Now that you've made it sticky, it'll only not be reported if client code never checks the future.

Perhaps it would have been simpler to resolve the TODO and throw directly. Client code would get the exception, and if it tried an auth() again all the same, so be it. But I'm also fine with your approach.

**Comment by lgoldstein** [comment]
>>  But I have no use case for calling auth() more than once on a session, and it's unclear to me what that would even mean. 

I  have thought of something (a bit out there, but I believe in the realm of what is allowed in SSH):

* Use tries password #1
* If  fails, then password #2,
* etc., etc.

>> Perhaps it would have been simpler to resolve the TODO and throw directly.

I'll consider it (in addition though to reporting the exception via the future)

**Comment by gnodet** [comment]
What about using something like the following instead of multiple fields to store the exception ?
```
diff --git a/sshd-core/src/main/java/org/apache/sshd/client/session/ClientSessionImpl.java b/sshd-core/src/main/java/org/apache/sshd/client/session/ClientSessionImpl.java
index 21e0631d..4037f23a 100644
--- a/sshd-core/src/main/java/org/apache/sshd/client/session/ClientSessionImpl.java
+++ b/sshd-core/src/main/java/org/apache/sshd/client/session/ClientSessionImpl.java
@@ -86,7 +86,6 @@ public class ClientSessionImpl extends AbstractClientSession {
         }

         authFuture = new DefaultAuthFuture(ioSession.getRemoteAddress(), futureLock);
-        authFuture.setAuthed(false);

         signalSessionCreated(ioSession);

@@ -124,9 +123,17 @@ public class ClientSessionImpl extends AbstractClientSession {
         ClientUserAuthService authService = getUserAuthService();
         synchronized (sessionLock) {
             String serviceName = nextServiceName();
-            authFuture = ValidateUtils.checkNotNull(
+            AuthFuture future = ValidateUtils.checkNotNull(
                     authService.auth(serviceName), "No auth future generated by service=%s", serviceName);
-            return authFuture;
+            future.addListener(f -> {
+                Throwable e = f.getException();
+                if (e != null) {
+                    this.authFuture.setException(e);
+                } else {
+                    this.authFuture.setAuthed(f.isSuccess());
+                }
+            });
+            return this.authFuture;
         }
     }
```

**Comment by lgoldstein** [comment]
Seems to work (need to think about it some more), but then the user can call `auth()` only once since the `authFuture` value will have been set, Is this something that violates SSH in some way ? I don't think SSH prohibits multiple attempts to authenticate...

**Comment by gnodet** [comment]
@lgoldstein you're right, there are a few tests that do use multiple calls to `auth()`.

**Comment by tomaswolf** [comment]
Didn't try, but this looks like it has the same problem again: if there's an exception before auth() it may be lost. Also authFuture is null initially, so more adaptations would be needed.

**Comment by gnodet** [comment]
Ok, I don't have much time to look at that, so feel free to go ahead with the PR.

**Pull request closed without merging**